runtime/drivers/druid/druidsqldriver/druid_api_sql_driver.go

```diff
@@ -40,193 +55,285 @@ type sqlConnection struct {
-					}
+
+				if errInvalidProtocol.MatchString(v.Error()) {
+					resp.Body.Close()
+					return nil, retrier.Fail, v
 				}
-			case "TIMESTAMP":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case string:
-						t, err := time.Parse(time.RFC3339, v)
-						if err != nil {
-							return nil, err
-						}
-						return t, nil
-					default:
-						return v, nil
-					}
+
+				if errTLSCert.MatchString(v.Error()) {
+					resp.Body.Close()
+					return nil, retrier.Fail, v
 				}
-			case "ARRAY":
-				transformers[i] = func(v any) (any, error) {
-					var l []any
-					err := json.Unmarshal([]byte(v.(string)), &l)
-					if err != nil {
-						return nil, err
-					}
-					return l, nil
+
+				// nolint:errorlint // there's no wrapping
+				if _, ok := v.Err.(x509.UnknownAuthorityError); ok {
+					resp.Body.Close()
+					return nil, retrier.Fail, v
 				}
-			case "OTHER":
-				transformers[i] = func(v any) (any, error) {
-					var l map[string]any
-					err := json.Unmarshal([]byte(v.(string)), &l)
-					if err != nil {
-						return nil, err
-					}
-					return l, nil
+			}
+
+			resp.Body.Close()
+			return nil, retrier.Retry, err
+		}
+
+		if resp.StatusCode == http.StatusTooManyRequests {
+			resp.Body.Close()
+			return nil, retrier.Retry, fmt.Errorf("Too many requests")
+		}
+
+		// Druid sends well-formed response for 200, 400 and 500 status codes, for others use this
+		// ref - https://druid.apache.org/docs/latest/api-reference/sql-api/#responses
+		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusBadRequest && resp.StatusCode != http.StatusInternalServerError {
+			resp.Body.Close()
+			return nil, retrier.Fail, fmt.Errorf("unexpected status code: %d, status: %s", resp.StatusCode, resp.Status)
+		}
+
+		dec := json.NewDecoder(resp.Body)
+
+		var obj any
+		err = dec.Decode(&obj)
+		if err != nil {
+			resp.Body.Close()
+			return nil, retrier.Fail, err
+		}
+		switch v := obj.(type) {
+		case map[string]any:
+			resp.Body.Close()
+			a := retrier.Fail
+			if em, ok := v["errorMessage"].(string); ok {
+				if errCoordinatorDown.MatchString(em) || errBrokerDown.MatchString(em) {
+					a = retrier.Retry
+				} else if errNoObject.MatchString(em) {
+					// if a table doesn't exist then it can be a restarting Coordinator
+					// note: there's still can be a restarting historical node that cannot be identifed by error messages
+					a = retrier.AdditionalCheck
 				}
 			}
+			// example:
+			// 500: Unable to parse the SQL
+			return nil, a, fmt.Errorf("%v", obj)
+		case []any:
+			columns := toStringArray(v)
+			err = dec.Decode(&obj)
+			if err != nil {
+				resp.Body.Close()
+				return nil, retrier.Fail, err
+			}
+
+			types := toStringArray(obj.([]any))
+
+			transformers := make([]func(any) (any, error), len(columns))
+			for i, c := range types {
+				transformers[i] = createTransformer(c)
+			}
+
+			druidRows := &druidRows{
+				closer:        resp.Body,
```

Comment:
Since `dec.Decode` has already been called on a JSON decoder that uses the body, won't it be fully consumed at this point? So it can just be closed immediately.

Reply:
> Decode reads the next JSON-encoded value from its input and stores it in the value pointed to by v.

It's NDJSON format, ie:
```
{"name": "Leandro","lastName": "Shokida"}
{"name": "Mariano","lastName": "De Achaval"}
```

Reply:
Got it, makes sense then